lightning/src/ln/peers/handshake/states.rs

```diff
@@ -1,30 +1,858 @@
-use ln::peers::handshake::hash::HandshakeHash;
+use bitcoin::secp256k1;
+
+use bitcoin::hashes::{Hash, HashEngine};
+use bitcoin::hashes::sha256::Hash as Sha256;
 use bitcoin::secp256k1::{SecretKey, PublicKey};
 
-pub enum HandshakeState {
-	Uninitiated,
-	AwaitingActOne(ActOneExpectation),
-	AwaitingActTwo(ActTwoExpectation),
-	AwaitingActThree(ActThreeExpectation),
-	Complete,
+use ln::peers::{chacha, hkdf};
+use ln::peers::conduit::{Conduit, SymmetricKey};
+
+const ACT_ONE_TWO_LENGTH: usize = 50;
+const ACT_THREE_LENGTH: usize = 66;
+
+type ChainingKey = [u8; 32];
+
+// Generate a SHA-256 hash from one or more elements concatenated together
+macro_rules! concat_then_sha256 {
```

Comment:
this is just personal opinion, but I'm rather partial to the previous hash instance approach over a macro for the sole reason of better legibility

Reply:
The one other benefit from legibility was standardizing the concat-before-sha pattern. Some of the code used a separate Vec<u8> to do the concat and passed it into a single `input()` and other parts of the code used 2x `input()` calls. This way, they all use the same path and it has the nice benefit of being able to read against the RFC text.

With that said, I completely understand that macros can get out of hand (see the same pattern in the hkdf refactor) and it might just be a good policy decision to use the extra 3 lines per invocation or split out a real function to do the work to minimize future errors.

If we move away from this macro I would advocate for standardizing on `Sha256::hash()` for the single input case instead of the previous 3-line pattern to clean it up a bit.

```
fn initialize_state(public_key: &PublicKey) -> (HandshakeHash, [u8; 32]) {
	let protocol_name = b"Noise_XK_secp256k1_ChaChaPoly_SHA256";
	let prologue = b"lightning";

	let mut sha = Sha256::engine();
	sha.input(protocol_name);
	let chaining_key = Sha256::from_engine(sha).into_inner();

	let mut initial_hash_preimage = chaining_key.to_vec();
	initial_hash_preimage.extend_from_slice(prologue.as_ref());

	let mut hash = HandshakeHash::new(initial_hash_preimage.as_slice());
	hash.update(&public_key.serialize());

	(hash, chaining_key)
}
```